src/cargo/core/registry.rs

```diff
@@ -63,6 +64,12 @@ pub trait Registry {
 
     /// Block until all outstanding [`Poll::Pending`] requests are [`Poll::Ready`].
     fn block_until_ready(&mut self) -> CargoResult<()>;
+
+    /// Get version preferences.
+    fn version_prefs(&self) -> &VersionPreferences;
+
+    /// Set version preferences.
+    fn set_version_prefs(&mut self, version_prefs: VersionPreferences);
```

Comment:
For myself, I feel uncomfortable how much resolver logic lives in `PackageRegistry`, I feel even more uncomfortable adding `VersionPreferences` to it and to `trait Registry` as well.